services/telemetry_v2/telemetry.go

```diff
@@ -0,0 +1,306 @@
+	l          *logrus.Entry
+	start      time.Time
+	config     ServiceConfig
+	dsRegistry TelemetryDataSourceLocator
+	pmmVersion string
+}
+
+func (s *Service) LocateTelemetryDataSource(name string) (TelemetryDataSource, error) {
+	return s.dsRegistry.LocateTelemetryDataSource(name)
+}
+
+// check interfaces
+var (
+	_ TelemetryDataSourceLocator = (*Service)(nil)
+)
+
+// NewService creates a new service.
+func NewService(db *reform.DB, pmmVersion string, config ServiceConfig) (*Service, error) {
+	if config.SaasHostname == "" {
+		return nil, errors.New("empty host")
+	}
+
+	l := logrus.WithField("component", "telemetry_v2")
+
+	registry, err := NewDataSourceRegistry(config, l)
+	if err != nil {
+		return nil, err
+	}
+	s := &Service{
+		db:         db,
+		l:          l,
+		pmmVersion: pmmVersion,
+		start:      time.Now(),
+		config:     config,
+		dsRegistry: registry,
+	}
+
+	return s, nil
+}
+
+// Run start sending telemetry to SaaS.
+func (s *Service) Run(ctx context.Context) {
+	if !s.config.Enabled {
+		s.l.Warn("service is disabled, skip Run")
+		return
+	}
+
+	ticker := time.NewTicker(s.config.Reporting.Interval)
+	defer ticker.Stop()
+
+	doSend := func() {
+		report, err := s.prepareReport(ctx)
+		if err != nil {
+			s.l.Debugf("Failed to prepare report: %s.", err)
+			return
+		}
+
+		err = s.send(ctx, report)
+		if err != nil {
+			s.l.Debugf("Telemetry info not sent, due to error: %s.", err)
+			return
+		}
+		s.l.Debug("Telemetry info sent.")
+	}
+
+	if s.config.Reporting.SendOnStart {
+		s.l.Debug("Telemetry on start is enabled, sending...")
+		doSend()
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			doSend()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *Service) prepareReport(ctx context.Context) (*reporter.ReportRequest, error) {
+	current, err := s.makeMetric()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, telemetry := range s.config.telemetry {
+		ds, err := s.LocateTelemetryDataSource(telemetry.Source)
+		if err != nil {
+			s.l.Debugf("failed to lookup telemetry datasource for [%s]:[%s]", telemetry.Source, telemetry.Id)
+			continue
+		}
+		if !ds.Enabled() {
+			continue
+		}
+
+		//TODO: consider fetching metrics in parallel
+		metrics, err := ds.FetchMetrics(ctx, telemetry)
+		if err != nil {
+			s.l.Debugf("failed to extract metric from datasource for [%s]:[%s]: %v", telemetry.Source, telemetry.Id, err)
+			continue
+		}
+
+		current.Metrics = append(current.Metrics, metrics...)
+	}
+
+	return &reporter.ReportRequest{
+		Metrics: []*pmmv1.ServerMetric{current},
+	}, nil
+}
+
+func (s *Service) makeMetric() (*pmmv1.ServerMetric, error) {
+	var settings *models.Settings
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		var e error
+		if settings, e = models.GetSettings(tx); e != nil {
+			return e
+		}
+
+		if settings.Telemetry.Disabled {
+			return errors.New("disabled via settings")
+		}
+		if settings.Telemetry.UUID == "" {
+			settings.Telemetry.UUID, e = generateUUID()
```

Comment:
in case we enable both telemetries, what will happen?

Reply:
they are independent, it will be sent 2 time.

in telemeteryD we support V1 and V2, so it will be stored in old and new format 2 times

Reply:
but we might generate it twice and store it to DB twice and override each other.

Reply:
I see, we can have race with V1 and V2. As a consequence on the first run there is a risk of having V1 with one ID and V2 with another.
I think V1 or V2 can be enabled, but not both.

Reply:
Do we have check for safety from that?

Reply:
I removed V1, I think it is not needed.